lib/AST/ASTContext.cpp

```diff
@@ -4270,6 +4272,64 @@ DependentMemberType *DependentMemberType::get(Type base,
   return known;
 }
 
+OpaqueTypeArchetypeType *
+OpaqueTypeArchetypeType::get(OpaqueTypeDecl *Decl,
+                             SubstitutionMap Substitutions) {
+  llvm::FoldingSetNodeID id;
+  Profile(id, Decl, Substitutions);
+  
+  auto &ctx = Decl->getASTContext();
+
+  // An opaque type isn't contextually dependent like other archetypes, so
+  // by itself, it doesn't impose the "Has Archetype" recursive property,
+  // but the substituted types might.
+  RecursiveTypeProperties properties = {};
+  for (auto type : Substitutions.getReplacementTypes()) {
+    properties |= type->getRecursiveProperties();
+  }
+  
+  auto arena = getArena(properties);
+  
+  llvm::FoldingSet<OpaqueTypeArchetypeType> *set;
+  switch (arena) {
+  case AllocationArena::Permanent:
+    set = &ctx.getImpl().OpaqueArchetypes;
+    break;
+  case AllocationArena::ConstraintSolver:
+    set = &ctx.getImpl().SolverOpaqueArchetypes;
+      break;
+  }
+  
+  void *insertPos;
+  if (auto existing = set->FindNodeOrInsertPos(id, insertPos)) {
+    return existing;
+  }
+  
+  // Create a new opaque archetype.
+  // It lives in an environment in which the interface generic arguments of the
+  // decl have all been same-type-bound to the arguments from our substitution
+  // map.
+  GenericSignatureBuilder builder(ctx);
+  builder.addGenericSignature(Decl->getOpaqueInterfaceGenericSignature());
+  // Same-type-constrain the arguments in the outer signature to their
+  // replacements in the substitution map.
+  for (auto outerParam : Decl->getGenericSignature()->getGenericParams()) {
+    auto boundType = Type(outerParam).subst(Substitutions);
+    builder.addSameTypeRequirementToConcrete(Type(outerParam),
+       boundType,
+       GenericSignatureBuilder::FloatingRequirementSource::forAbstract(),
+       GenericSignatureBuilder::UnresolvedHandlingKind::GenerateConstraints,
+       [](Type, Type) { llvm_unreachable("error?"); });
+  }
+  
+  auto newOpaque = std::move(builder)
+    .computeOpaqueArchetype(Decl, Substitutions, properties, arena);
+  
+  set->InsertNode(newOpaque, insertPos);
```

Comment:
Please perform another `FindNodeOrInsertPos` here. Some other code might have perturbed the set, invalidating `insertPos`.